hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpc/EventloopFactory.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.tpc;
+
+import com.hazelcast.internal.util.ThreadAffinity;
+
+import java.util.concurrent.ThreadFactory;
+import java.util.function.Supplier;
+
+import static com.hazelcast.internal.tpc.util.Preconditions.checkNotNegative;
+import static com.hazelcast.internal.tpc.util.Preconditions.checkNotNull;
+import static com.hazelcast.internal.tpc.util.Preconditions.checkPositive;
+import static java.lang.System.getProperty;
+
+/**
+ * A factory for {@link Eventloop} instances.
+ */
+public abstract class EventloopFactory {
+    public static final String NAME_LOCAL_TASK_QUEUE_CAPACITY = "hazelcast.tpc.localTaskQueue.capacity";
+    public static final String NAME_CONCURRENT_TASK_QUEUE_CAPACITY = "hazelcast.tpc.concurrentTaskQueue.capacity";
+    public static final String NAME_SCHEDULED_TASK_QUEUE_CAPACITY = "hazelcast.tpc.scheduledTaskQueue.capacity";
+    public static final String NAME_BATCH_SIZE = "hazelcast.tpc.batch.size";
+    public static final String NAME_CLOCK_REFRESH_PERIOD = "hazelcast.tpc.clock.refreshPeriod";
+    public static final String NAME_EVENTLOOP_SPIN = "hazelcast.tpc.eventloop.spin";
+
+    private static final int DEFAULT_LOCAL_QUEUE_CAPACITY = 65536;
+    private static final int DEFAULT_CONCURRENT_QUEUE_CAPACITY = 65536;
+    private static final int DEFAULT_SCHEDULED_TASK_QUEUE_CAPACITY = 4096;
+    private static final int DEFAULT_BATCH_SIZE = 64;
+    private static final int DEFAULT_CLOCK_REFRESH_INTERVAL = 16;
+    private static final boolean DEFAULT_SPIN = false;
+
+    protected final EventloopType type;
+    Supplier<Scheduler> schedulerSupplier = NopScheduler::new;
+    Supplier<String> threadNameSupplier;
+    ThreadAffinity threadAffinity = ThreadAffinity.newSystemThreadAffinity("hazelcast.tpc.eventloop.affinity");
+
+    ThreadFactory threadFactory = Thread::new;
```

Comment:
I don't really like the fact, that Eventloop and EventloopFactory have circular dependencies. It looks liek EventloopFactory is used not only to create eventloops with given config, but also is responsible for thread-stuff - schedulerSupplier, threadNameSupplier, threadAffinity, threadFactory - all of those are not used by `create`, only by Eventloop, so I think it's better to extract to separate class

Reply:
The EventloopFactory contains the dependencies needed to create the Eventloop and the Eventloop reads these properties from the factory. I really don't see a reason to split them. This functionality is not intended to be used independently of each other.

Reply:
So then those properties should be passed to the Eventloop - having circular dependency is a big risk of future technical debt when code grows and make it non-such-dependent will be much harder

Reply:
Classes with a large number of arguments are a PITA and creating a bosus 'argument passing' setting object just leads to writing the same field multiple times.

And as I already indicated, cyclic dependency in this case is a nonissue since the classes belong to each other. They are part of the same package and are not intended to be used without each other.

Reply:
Have a look at the alto-experimental branch and try to fix this problem. Then you have 2 'configuration' subtypes to deal with and not just 1.

If you know of a solution that doesn't lead to a super fat constructor or repeating all fields of the object, then I'm all ears.

Reply:
I renamed the EventloopFactory to EventloopBuilder (still with the cycle). Builder is more appropriate name for this class than Factory.

Reply:
From the author of effective Java:

https://blogs.oracle.com/javamagazine/post/exploring-joshua-blochs-builder-design-pattern-in-java